hack/generator/pkg/codegen/pipeline_collapse_cross_group_refs.go

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+	"github.com/pkg/errors"
+)
+
+// collapseCrossGroupReferences finds and removes references between API groups. This isn't particularly common
+// but does occur in a few instances, for example from Microsoft.Compute -> Microsoft.Compute.Extensions.
+func collapseCrossGroupReferences() PipelineStage {
+	return MakePipelineStage(
+		"collapseCrossGroupReferences",
+		"Finds and removes cross group references",
+		func(ctx context.Context, types astmodel.Types) (astmodel.Types, error) {
+			resources := astmodel.CollectResourceDefinitions(types)
```

Comment:
Is this deterministing in sort order of the results? If not, or if the return type is a map, then sometimes we might merge `Microsoft.Compute` into `Microsoft.Compute.Extensions` intead of the other way around.

Reply:
A good point - will improve. Although in the case of your example Microsoft.Compute always references Microsoft.Compute.Extensions, which doesn't itself reference Microsoft.Compute I don't believe (there's not a cycle). But there could be one so I'll ensure we handle it.

Reply:
Oh wait actually what we're doing now is ok: If there _were_ a cycle then the stuff in Microsoft.Compute which Microsoft.Compute.Extensions referenced would get copied into Microsoft.Compute.Extensions, and the stuff in Microsoft.Compute.Extensions which was referenced by Microsoft.Compute will be moved into Microsoft.Compute.

So the net is in a full cycle that both packages will contain everything in the other package - which is the desired behavior so that we're not crossing groups for references.

Reply:
Makes sense.

Reply:
Resolving this as there was agreement that it is correct over Teams.